### DISCUSSION TITLE
Are DataTypes needed?

### DISCUSSION THREAD

**Discussion by mtf90** [LearnLib/ralib#73] [AutomataLib Q & A] [answered]
Do we actually need `DataType` as a standalone entity, or could the model be simplified by using the respective `Class<?>` as type constraint directly? The only place where I could see it being useful is in the teachers definitions (`Map<DataType, Theory>`). Having it as an explicit entity would allow one to use different theories for same-typed data types simultaneously (e.g., an integer `EqualityTheory` for one parameter type and an integer `InequalityTheory` for another parameter type). Is this something that should be supported?

**Comment by FredrikTaquist** [comment, marked as answer]
The `DataType` entity allows us to distinguish between types that happen to have the same `Class<?>`, but should not be treated as the same data type. For example, say that you have a `userID` and `PIN` data type, which can both be represented as integers, but are clearly not the same type.

Also, as you suggested the question, having a separate `DataType` allows us to have different theories for the different data types, even if the underlying `Class<?>` is the same. Since we plan to support more complex theories, such as learning with successors (guards of the form `s2 == s1 + 1`) it is useful to have this functionality.

**Reply by mtf90** [reply]
Ok, then it makes sense to leave the concept as-is.
